### PR TITLE
feat: download page notices and download all button

### PR DESCRIPTION
**Pull request by tomwisecodes**
## Description

Partial implementation of 
https://figma.com/design/60B9J4YPWyzytpFNhyoQgx/Downloads?node-id=70-7129&node-type=canvas&t=BHMZ8hdO8TOm4HeW-0

### CODE REVIEW

File: packages/api/src/router/exports.ts
Comment (tomwisecodes) [inline packages/api/src/router/exports.ts:1]: This is a large diff as it moves the functional logic of each of the exports to its own folder

File: packages/api/src/export/exportAdditionalMaterialsDoc.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import { SignedInAuthObject } from "@clerk/backend/internal";
```
Comment (tomwisecodes) [inline packages/api/src/export/exportAdditionalMaterialsDoc.ts:1]: These files could be refactored as a follow up to remove some common components

File: apps/nextjs/src/middlewares/csp.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ const sentryEnv = process.env.NEXT_PUBLIC_SENTRY_ENV;
 const sentryRelease = process.env.NEXT_PUBLIC_APP_VERSION;
```
Comment (tomwisecodes) [inline apps/nextjs/src/middlewares/csp.ts:4]: have had to play with csp to get the api route js to work

File: apps/nextjs/src/app/api/aila-download-all/route.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+Loading...
+| 
+Please wait while we prepare your files for download. This can take up to 1 minute.
+
+<script nonce="${nonce}">
+  const spinnerChars = ['|', '/', '-', '\\\\'];
+  let spinnerIndex = 0;
+  
+  setInterval(() => {
+    document.querySelector('pre').textContent = 'Loading...\\n' + spinnerChars[spinnerIndex] + '\\nPlease wait while we prepare your files for download. This can take up to 1 minute.';
```
Comment (mantagen) [inline apps/nextjs/src/app/api/aila-download-all/route.ts:107]: 😵‍💫

Comment (mantagen) [reply apps/nextjs/src/app/api/aila-download-all/route.ts:107]: I don't think we need this .. i think we can just have [`a href="api_download_route" download /`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download)`

File: apps/nextjs/src/app/api/aila-download-all/route.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      passThrough.destroy();
+    },
+  });
+}
+
+async function getHandler(req: Request): Promise<Response> {
+  const { searchParams } = new URL(req.url);
+  const fileIdsParam = searchParams.get("fileIds");
+  const lessonTitle = searchParams.get("lessonTitle") ?? "";
+  const { userId }: { userId: string | null } = auth();
```
Comment (mantagen) [inline apps/nextjs/src/app/api/aila-download-all/route.ts:75]: do we need that type? isn't it already coming from auth()

File: apps/nextjs/src/app/api/aila-download-all/route.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    cancel() {
+      passThrough.destroy();
+    },
+  });
+}
+
+async function getHandler(req: Request): Promise<Response> {
+  const { searchParams } = new URL(req.url);
+  const fileIdsParam = searchParams.get("fileIds");
+  const lessonTitle = searchParams.get("lessonTitle") ?? "";
```
Comment (mantagen) [inline apps/nextjs/src/app/api/aila-download-all/route.ts:74]: can we error if no lesson title?

File: apps/nextjs/src/app/api/aila-download-all/route.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        const res = await downloadDriveFile({ fileId, ext });
+
+        if ("error" in res) {
+          Sentry.captureException(res.error);
+          continue;
+        }
+
+        const { data } = res;
+
+        const shortId = lessonExport.id.slice(-8);
```
Comment (mantagen) [inline apps/nextjs/src/app/api/aila-download-all/route.ts:175]: Why we slicing from the end not beginning?

File: apps/nextjs/src/components/AppComponents/Chat/chat-left-hand-side.tsx [hunk 7]
```diff
@@ -15,6 +17,7 @@ type ChatLeftHandSideProps = {
   isDemoLocked: boolean;
   showLessonMobile: boolean;
   setShowLessonMobile: (value: boolean) => void;
+  demo: DemoContextProps;
```
Comment (mantagen) [inline apps/nextjs/src/components/AppComponents/Chat/chat-left-hand-side.tsx:20]: What's the demo thing being used for here?

File: apps/nextjs/src/components/AppComponents/download/DownloadAllButton.tsx [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        exitQuiz: string | undefined;
+      }
+    | {
+        error: unknown;
+        message: string;
+      }
+    | undefined;
+  exportsType: ExportsType;
+  "data-testid"?: string;
+}) => {
```
Comment (mantagen) [inline apps/nextjs/src/components/AppComponents/download/DownloadAllButton.tsx:64]: Could these props be their own types so it's a bit clearer?

File: apps/nextjs/src/components/AppComponents/download/DownloadAllButton.tsx [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+              Download all resources in pdf and docx/pptx
+            </span>
+            <span className="text-left opacity-80">All sections</span>
+          </div>
+        </Link>
+        <span className="my-12 h-[2px] w-full bg-black opacity-15" />
+
+        <button
+          className="flex w-full items-center justify-start gap-15 hover:underline"
+          onClick={async () => {
```
Comment (mantagen) [inline apps/nextjs/src/components/AppComponents/download/DownloadAllButton.tsx:121]: Could this click handler not be inlined?

File: apps/nextjs/src/components/AppComponents/download/DownloadAllButton.tsx [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        {icon === "slides" && <SlidesIcon />}
+        {icon === "lesson" && <LessonIcon />}
+        {icon === "quiz" && <QuizIcon />}
+      </Box>
+    );
+  }
+  return null;
+}
+
+function handleSendEmailIcon({
```
Comment (mantagen) [inline apps/nextjs/src/components/AppComponents/download/DownloadAllButton.tsx:239]: This looks kind of like a function component i.e. `<SendEmailIcon isError={} ... />`

File: apps/nextjs/src/components/ExportsDialogs/useExportAllLessonAssets.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  onStart,
+  lesson,
+  chatId,
+  messageId,
+  active,
+}: {
+  onStart: () => void;
+  lesson: LessonDeepPartial;
+  chatId: string;
+  messageId: number;
```
Comment (mantagen) [inline apps/nextjs/src/components/ExportsDialogs/useExportAllLessonAssets.ts:24]: Will need to be a string now that the message id work has gone in!

File: apps/nextjs/src/middlewares/csp.ts [hunk 12]
```diff
@@ -186,26 +186,18 @@ const buildCspHeaders = (nonce: string) => {
...
   res: NextResponse | Response,
   req: Request,
   newHeaders: Record<string, string>,
 ) => {
   if (!res.headers.get(OVERRIDE_HEADERS)) {
-    // Emulate a user setting overrides by explicitly adding the required nextjs headers
-    // https://github.com/vercel/next.js/pull/41380
-    // @ts-expect-error Argument of type 'string[]' is not assignable to parameter of type 'string'.ts(2345)
-    res.headers.set(OVERRIDE_HEADERS, [...req.headers.keys()]);
+    res.headers.set(OVERRIDE_HEADERS, [...req.headers.keys()].join(","));
```
Comment (mantagen) [inline apps/nextjs/src/middlewares/csp.ts:195]: What's the change here about?

File: packages/api/src/router/exports.ts [hunk 13]
```diff
@@ -1074,152 +752,173 @@ export const exportsRouter = router({
...
+          exitQuizLink,
+          additionalMaterialsLink,
+        } = input;
+
+        if (!userEmail) {
+          console.error("User email not found");
+          return false;
+        }
+
+        const emailSent = await sendEmail({
```
Comment (mantagen) [inline packages/api/src/router/exports.ts:891]: Were we sending this before?

File: apps/nextjs/src/middlewares/csp.ts (hunk 12 above)
Comment (tomwisecodes) [reply apps/nextjs/src/middlewares/csp.ts:195]: fixes a type issue that came about

File: apps/nextjs/src/components/AppComponents/Chat/chat-left-hand-side.tsx (hunk 7 above)
Comment (tomwisecodes) [reply apps/nextjs/src/components/AppComponents/Chat/chat-left-hand-side.tsx:20]: There is an inchat button that needs blocking

File: apps/nextjs/src/app/api/aila-download-all/route.ts (hunk 3 above)
Comment (tomwisecodes) [reply apps/nextjs/src/app/api/aila-download-all/route.ts:107]: can chat about this on slack but that will lead the user to a blank page which they have to wait at for a min so they need some sort of indication that they have to wait.

**Pull request closed without merging**